Seed in the skeleton of ML-DSA based on current prototyping

bartonjs opened the pull request:
This seeds a lot of untested boilerplate code as non-public types in System.Security.Cryptography.

When we start work on ML-DSA, we can take this, make MLDsa and MLDsaAlgorithm public, start to write tests, and wire up an implementation. We can also take this and clone+customize it for the other PQC algorithms.

It does not define all span-vs-array overloads for all spanified members, just one candidate per method group.

As the PQC experiment needs to make it further before we can commit to the shape, it is both starting without API Review, and pre-emptively applying the Experimental attribute.

bartonjs committed:
Seed in the skeleton of ML-DSA based on current prototyping

bartonjs committed:
Suppress IDE0060 in the stub file

bartonjs commented on src/libraries/Common/src/System/Security/Cryptography/KeyFormatHelper.Encrypted.cs:
```diff
@@ -272,6 +272,68 @@ internal static ArraySegment<byte> DecryptPkcs8(
+        {
+            fixed (byte* pointer = source)
+            {
+                using (PointerMemoryManager<byte> manager = new(pointer, source.Length))
```
Yeah, there's a fair amount of "can we move this down?" that I had while stubbing this out.  Some of it, like doing a PKCS8 encryption envelope over bytes instead of AsnWriter, just looked too big for this particular seed-things change.

I'll probably be looking at changing how we do storage in the Asn types in the next couple of weeks, to see if we can get rid of the Memory requirement that I forced upon us all those years ago, (and we can get span-friendlier)

bartonjs commented on src/libraries/Common/src/System/Security/Cryptography/MLDsa.cs:
```diff
@@ -0,0 +1,1803 @@
+        {
+        }
+
+        protected void ThrowIfDisposed()
```
I was thinking part of the API surface, so that it didn't need to be repeated at every level (unlike our current pattern).  If (e.g. MLDsaKeyVault wanted some sort of extra `ExportToSomeOtherKeyVault(??? otherKeyVault)`).

If we don't like it, we can cut it.  If API Review doesn't like it, we might be forced to cut it.  But the current proposal was intentional.

bartonjs commented on src/libraries/Common/src/System/Security/Cryptography/MLDsa.cs:
```diff
@@ -0,0 +1,1803 @@
+                throw new ArgumentException(nameof(destination), SR.Argument_DestinationTooShort);
+            }
+
+            ExportMLDsaPublicKeyCore(destination.Slice(0, PublicKeySizeInBytes));
```
Yeah, it's interesting because we know exactly how big the FIPS 204 format is per parameter set.  I had a brief "if it's void someone could leave it blank by accident...", then decided "eh, they should have tests".

But since part of the purpose of the template pattern is to avoid needing to do argument validation; the caller always passing an exact-sized buffer seemed right... at which point the return was silly (all the overrider can do is be wrong)

bartonjs commented on src/libraries/Common/src/System/Security/Cryptography/MLDsa.cs:
```diff
@@ -0,0 +1,1803 @@
+        {
+            ThrowIfNotSupported();
+
+            // TODO, probably in PemKeyHelper.
```
When I noticed (right before putting up the PR) that the IImportExportShape didn't have PEM, I started by leaving the Try off and just leaving a comment as to how it might work, considering that it has three failure modes: 1) no viable PEM was found, 2) encrypted private key was found, 3) viable PEM was found, but didn't indicate the algorithm.

If we want to say that static ImportFromPem works like RSA's instance ImportFromPem, and it's "there better only be one semi-viable one"... that seems sound. So then the Try returns false for "no viable input, or the only viable input wasn't ML-DSA", and throws for corrupt data or too many viable options.  Which feels weird.  So maybe we don't want the Try at all.  Which then suggests maybe we don't want the other Try methods at all.

The Try for SPKI/P8 are mostly for an audience like SignTool.  "Sign the exe using this key" (or verify).  If it also has a cert (like Authenticode) then there's no guessing; but if it's something purely key-based (like COSE) then there's more guessing involved.  Since it's niche, maybe the best answer is to just cut all the Try-Imports until there's a better demonstrated need.

The main reason I think people do cascading "try-catch RSA, try-catch ECDSA, ..." is because they're all AsymmetricAlgorithm; and once we take that crutch away they need to think more about how their data flows... so maybe it's not necessary.

bartonjs commented on src/libraries/Common/src/System/Security/Cryptography/MLDsaAlgorithm.cs:
```diff
@@ -0,0 +1,126 @@
+    ///   Represents a specific algorithm within the ML-DSA family.
+    /// </summary>
+    [Experimental(Experimentals.PostQuantumCryptographyDiagId)]
+    internal struct MLDsaAlgorithm : IEquatable<MLDsaAlgorithm>
```
I think the main reason we use structs for strongly typed strings is so you don't feel compelled to optimize caching them when your data comes externally.

We kind of don't want data coming externally here... (a 3rd party extension can't meaningfully say they implement some ML-DSA variant the base class doesn't, since it'll throw when failing to resolve the parameter set)... so maybe we want a numeric enum, or a class string-enum.

vcsjones replied:
My thought would be make it a `sealed class`.

I agree that it is kind of nonsense for external data here - `ParameterSetInfo` is not going to know what to do with anything other than what we already know about. If we have a class, then

1. `default` is null, and the documentation and runtime checks are `ThrowIfNull`
2. We could have no public constructors, or, we can validate in the constructor names we know about, so that way everyone that consumes `MLDsaAlgorithm` doesn't have to check if the name is something we know about.

bartonjs replied:
Would you give the same treatment to the spiritual equivalent for all of ML-KEM, Composite ML-DSA, SLH-DSA, and Composite SLH-DSA?

Once it's locking down like this, I feel like we'd just merge the ParameterSetInfo data into it (so long as it's `internal` data it only matters to reflection where it lives)

vcsjones replied:
> Would you give the same treatment to the spiritual equivalent for all of ML-KEM, Composite ML-DSA, SLH-DSA, and Composite SLH-DSA?

Yes.

> Once it's locking down like this, I feel like we'd just merge the ParameterSetInfo data into it

That actually seems quite reasonable now. It can all be internal.

bartonjs committed:
Address feedback

bartonjs committed:
Add missing registration of the experimental diagnostic ID

bartonjs committed:
Remove the TryImport methods

bartonjs committed:
Leave TODO for Pre-Hash variants

bartonjs committed:
Minor cleanup

bartonjs committed:
Change MLDsaAlgorithm to a class

stephentoub commented on src/libraries/Common/src/System/Security/Cryptography/IImportExportShape.cs:
```diff
@@ -0,0 +1,37 @@
+
+namespace System.Security.Cryptography
+{
+#if DESIGNTIMEINTERFACES
```
What's the purpose of this ifdef? This file appears to only be consumed into the S.S.Cryptography.csproj, which sets this.

bartonjs replied:
It’ll also be consumed in Microsoft.Bcl.Cryptography as the project gets further along

stephentoub commented on src/libraries/Common/src/System/Security/Cryptography/KeyFormatHelper.Encrypted.cs:
```diff
@@ -272,6 +274,68 @@ internal static ArraySegment<byte> DecryptPkcs8(
+                {
+                    ArraySegment<byte> decrypted = KeyFormatHelper.DecryptPkcs8(password, manager.Memory, out bytesRead);
+                    AsnValueReader reader = new(decrypted, AsnEncodingRules.BER);
+                    reader.ReadEncodedValue();
```
These lines can't throw, or we don't want to return decrypted to the pool if they do?

bartonjs replied:
Fixed.  I think I based this on some other code where it had run through Decode to know it was structurally valid, and I just needed to know if it had trailing data.  But since this hasn't, those should be in the try.

stephentoub commented on src/libraries/Common/src/System/Security/Cryptography/MLDsa.cs:
```diff
@@ -0,0 +1,1451 @@
+            while (!TryExportPkcs8PrivateKey(rented, out written))
+            {
+                CryptoPool.Return(rented, 0);
+                rented = CryptoPool.Rent(rented.Length * 2);
```
I realize this is safe, but seeing `rented` used after it's returned makes me wince.

bartonjs replied:
I had the same feeling when I wrote it... so now it doesn't do that any more.

(Every now and then I try to write smaller code, see? :smile:)

stephentoub commented on src/libraries/Common/src/System/Security/Cryptography/MLDsa.cs:
```diff
@@ -0,0 +1,1451 @@
+        private static void ThrowAlgorithmUnknown(AsnWriter encodedId)
+        {
+#if NET9_0_OR_GREATER
+            throw encodedId.Encode(static encoded =>
```
What's the benefit of this .NET 9 path as compared to the other one?

bartonjs committed:
Respond to feedback

bartonjs committed:
Seal MLDsaAlgorithm

Both because it's currently an internal type (and failing the build),
but also because it should be sealed as public API.

vcsjones reviewed (approved):
Looks good to me, API-shape wise.

bartonjs commented:
/ba-g Uncalled code, and the test are only from "'findstr' is not recognized as an internal or external command,"
